Fixed Issue #1508

MohamedLebda opened the pull request:
![bug #1508](https://user-images.githubusercontent.com/98412370/220509557-54618778-b2c2-4c12-9da9-d204047bb6ac.PNG)

![bug 1508](https://user-images.githubusercontent.com/98412370/220509767-a56a49b9-4134-4696-8d78-d5d5ce2e82d2.PNG)
 as we can see here when we choose autoplay off and playlist auto plays off the pause of the video become dependent on ImprovedTube.allow_autoplay once it equals false it stops the video and only plays again if the value of ImprovedTube.allow_autoplay changes to true and that happens, when we start the video again the reason it becomes false, is when the ImprovedTube.video_url !== location. href but in ImprovedTube.initPlayer  in function.js the  ImprovedTube.video_url = location.href  and that function fires when the event listener of yt-navigate-finish' fire  but 
The "yt-navigate-finish" event listener in the YouTube website's scripts will not work in the following situations:

When the YouTube website is not loaded: If the user is not currently on the YouTube website, the event listener will not be triggered.

When the page does not finish loading: If there is an error or some other issue that prevents the page from fully loading, the "yt-navigate-finish" event will not be triggered.

When the listener is not properly registered: If the event listener is not registered correctly or is removed before the page finishes loading, it will not be triggered.

When the listener conflicts with other scripts: If there are other scripts running on the page that conflict with the "yt-navigate-finish" event listener, it may not be triggered.

When the page is dynamically updated: If the page is dynamically updated using AJAX or other similar techniques, the "yt-navigate-finish" event may not be triggered as it only listens for full page loads.

In general, the "yt-navigate-finish" event listener is reliable and should work in most situations where a user navigates to a new page or video on the YouTube website. However, it is important to keep in mind the limitations listed above and to test your code thoroughly to ensure that it works as expected.

which means that in those cases ImprovedTube.video_url will never equal  location.href so it causes the bug of constant pause (freeze)

ImprovedTube commented:
> ImprovedTube.allow_autoplay once it equals false it stops the video and only plays again if the value of

>which means that in those cases ImprovedTube.video_url will never equal location.href so it causes the bug of constant pause (freeze)

  `ImprovedTube.allow_autoplay`     should be called    `ignore_autoplay_off`   

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/core.js#L45-L47

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/functions.js#L362-L388

#1456 : 

where's the ethernal loop? 

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/be5a182bfb41159850490e8dc83806e57931211b/content-scripts/website-context/functions.js#L215
this could exclude /videos$  /shorts$ /playlists$
(also there is a \/  too much /\/channel|\/user|\/c\/|\/@.*/.test(location.href))    )

---

 - autoplay prevention  only needs to happen within the first few seconds seconds a \<video\> is  loaded / active 

><img src="https://user-images.githubusercontent.com/98412370/220509767-a56a49b9-4134-4696-8d78-d5d5ce2e82d2.PNG" 

what if we remove this?  will emdedded autoplay=1  players make trouble?

---

> When the YouTube website is not loaded: If the user is not currently on the YouTube website, the event listener will not be triggered.

which?

ImprovedTube commented:
- autoplay prevention only needs to happen within the first few seconds seconds a \<video\> is  loaded / active  🤔

> ImprovedTube.allow_autoplay once it equals false it stops the video and only plays again if the value of

>which means that in those cases ImprovedTube.video_url will never equal location.href so it causes the bug of constant pause (freeze)

 `ImprovedTube.allow_autoplay` 
 should be called    `ignore_autoplayOFF_because_of_user_action`   

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/core.js#L45-L47

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/functions.js#L362-L388

----

#1456 : 

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/be5a182bfb41159850490e8dc83806e57931211b/content-scripts/website-context/functions.js#L215

 - this could exclude /videos$       (also there is a \/  too much /\/channel|\/user|\/c\/|\/@.*/.test(location.href))    )

---

> When the YouTube website is not loaded: If the user is not currently on the YouTube website, the event listener will not be triggered.

example? 

><img src="https://user-images.githubusercontent.com/98412370/220509767-a56a49b9-4134-4696-8d78-d5d5ce2e82d2.PNG" 

what if we remove this?  will emdedded autoplay=1  players make trouble?   #1227  #1065

ImprovedTube commented:
> ImprovedTube.allow_autoplay once it equals false it stops the video and only plays again if the value of

>which means that in those cases ImprovedTube.video_url will never equal location.href so it causes the bug of constant pause (freeze)

doesnt seem to cause an ethernal loop 🤔
`ImprovedTube.allow_autoplay`   should be called    `ignore_autoplayOFF_because_of_user_action`   

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/core.js#L45-L47

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/functions.js#L362-L388

><img src="https://user-images.githubusercontent.com/98412370/220509767-a56a49b9-4134-4696-8d78-d5d5ce2e82d2.PNG" 

what if we removed this?   

----

- autoplay prevention only needs to happen within the first few seconds seconds a \<video\> is  loaded / active  🤔

https://stackoverflow.com/questions/37706626/yt-getplayerstate-not-a-function

https://stackoverflow.com/questions/34077641/how-to-detect-page-navigation-on-youtube-and-modify-its-appearance-seamlessly

---

#1456 : 

https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/be5a182bfb41159850490e8dc83806e57931211b/content-scripts/website-context/functions.js#L215

 - this could exclude /videos$       (also there is a \/  too much /\/channel|\/user|\/c\/|\/@.*/.test(location.href))    )

---

 iframes  /emded/...&autoplay=1      #1227  #1065

MohamedLebda commented:
> doesn't seem to cause an eternal loop 🤔
it will because then ImprovedTube.allow_autoplay will always still be false because will always stay ImprovedTube.video_url !== location.href if "yt-navigate-finish" didn't work.
> what if we removed this?
that will cause the video to freeze (always in pause if (auto_play is off) because the ImprovedTube.allow_autoplay will forever be false

MohamedLebda commented:
> doesnt seem to cause an ethernal loop 🤔


it will because then ImprovedTube.allow_autoplay will always still be false because will always stay ImprovedTube.video_url !== location.href if "yt-navigate-finish" didn't work.


> what if we removed this?

that will cause the video to freeze (always in pause if (auto_play is off) because the ImprovedTube.allow_autoplay will forever be false

ImprovedTube commented:
> always

only onkeydow || onmousedown html5-main-video,   ytp-play-button

> i meant removing the whole 5 lines https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/f73e4180a816d87663e87601bba75a9cb799a39c/content-scripts/website-context/youtube-features/player.js#L10-L14

ImprovedTube commented:
> always

no, only if the user is active already while before autoplay
onkeydow || onmousedown html5-main-video,   ytp-play-button

>>> what if we removed this 

 i meant, please try to understand, what is the disanvantage when removing the 5 lines https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/f73e4180a816d87663e87601bba75a9cb799a39c/content-scripts/website-context/youtube-features/player.js#L10-L14

MohamedLebda commented:
> [ImprovedTube-for-YouTube/content-scripts/website-context/functions.js](https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/ba7c738ab1cc429899a8786465334347da567eda/content-scripts/website-context/functions.js#L362-L388)
> 
> Lines 362 to 388 in [ba7c738](/code-for-charity/ImprovedTube-for-YouTube/commit/ba7c738ab1cc429899a8786465334347da567eda)

sorry I thought you meant this.


> removing the 5 lines

if we removed the first line which is var video = ImprovedTube.elements.player; 
that will result in the video will never pause at the start even tho the auto-play is off because the video will not be identified  in 

   if (ImprovedTube.allow_autoplay === false && video.classList.contains('ad-showing') === false &&
        (
            (document.documentElement.dataset.pageType === "video" && ((location.href.indexOf('list=') === -1 && ImprovedTube.storage.player_autoplay === false) || (ImprovedTube.storage.playlist_autoplay === false && location.href.indexOf('list=') !== -1))) ||
            (document.documentElement.dataset.pageType === "channel" && ImprovedTube.storage.channel_trailer_autoplay === false)
        )
    ) {
        setTimeout(function () {
            video.pauseVideo();
        });
    }
};

and if we removed the next 4 lines in 

> [ImprovedTube-for-YouTube/content-scripts/website-context/youtube-features/player.js](https://github.com/code-for-charity/ImprovedTube-for-YouTube/blob/f73e4180a816d87663e87601bba75a9cb799a39c/content-scripts/website-context/youtube-features/player.js#L10-L14)
> 
> Lines 10 to 14 in [f73e418](/code-for-charity/ImprovedTube-for-YouTube/commit/f73e4180a816d87663e87601bba75a9cb799a39c)

that will result in( auto play off will not work when open a new video because the value of ImprovedTube.allow_autoplay will be true form when we opened a  new video
